Per-component themeing

lederer opened the pull request:
## Overview

Refactor theme system to support per-component themeing.

Replaced `defaultTheme.js` with a `theme` folder:

```
packages/core/theme/
├── components/
│   ├── foo/
│       └── index.js
│   ├── bar/
│       └── index.js
│   └── index.js
├── base.js
├── icons.js
└── index.js
```

What was `defaultTheme.js` is now `base.js` and, as before, contains component-agnostic theme settings organized into styled-system theme objects.

`icons.js` remains untouched. I updated the CLI to its new location.

`index.js` compiles and exports `base.js`, `icons.js`, and the component theme files as a single `theme` object, with per-component theme properties distributed across their styled-system theme objects. Any references in the component themes to keys in `base.js` get resolved along the way.

For example, this `base.js`...

```js
// base.js
{
  space: [
    "0",
    "0.8rem",
    "1.6rem",
    ...
  ],
  colors: {
    white: "#ffffff",
    black: "#000000",
    primary: "#738FFC",
    secondary: "#2FC1D0",
    ...
  },
  ...
}
```

... and these component themes ...

```js
// components/foo.js

{
  space: {
    px: 2,
    py: 1
  },
  colors: {
    color: "primary",
    bg: "white"
  },
  ...
}
```

```js
// components/bar.js

{
  space: {
    m: 1,
    p: 1
  },
  colors: {
    color: "black",
    highlight: "secondary"
  },
  ...
}
```
result in this `theme` object:

```js
{
  space: {
    0: 0,
    1: "0.8rem",
    2: "1.6rem",
    ...
    foo: {
      px: "1.6rem",
      py: "0.8rem"
    },
    bar: {
      m: "0.8rem",
      p: "0.8rem"
    }
  },
  colors: {
    white: "#ffffff",
    black: "#000000",
    primary: "#738FFC",
    secondary: "#2FC1D0",
    ...
    foo: {
      color: "#738FFC",
      bg: "#ffffff"
    },
    bar: {
      color: "#000000",
      highlight: "#2FC1D0"
    }
  },
  ...
}
```


### Checklist

- [x] Relevant documentation pages have been created or updated
- [ ] Description of PR is in an appropriate section of the changelog and grouped with similar changes if possible

Are there any of the following in this PR?

- [ ] Changes to the prop names or types of existing components
- [ ] Changes in intended behavior of existing component props
- [x] Changes in the theme file's structure
- [x] A required version bump to a non-dev dependency of the project

If one of the above is checked...

- [x] information or guidance around needed changes for consumers of this library has been added to the changelog under `Upgrade Instructions`


### Notes

I had to bump `styled-system` to `3.2.1`.

The CLI needs to be updated to automatically generate (or destroy) the corresponding subfolder of `theme/components/`, and update `theme/components/index.js`, when `blaster g` (or `blaster d`) is run.


## Testing Instructions

 * Fiddle with values in `base.js`, make sure changes propagate to components that refer to those keys in their component themes.
 * Fiddle with values in a component's theme files. Use keys from `base.js` as well as valid strings or numbers that the corresponding props expect.

lederer committed:
Bump styled-system version, for nested objects in theme.space

lederer committed:
Per-component themeing. Starting with badge, breadcrumbs.

lederer committed:
Port existing components to per-component themeing

lederer committed:
Update README to include per-component themeing instructions

lederer committed:
CHANGELOG

lederer commented:
@alkamin I think this PR should be blocked by https://github.com/raster-foundry/blasterjs/issues/53#issuecomment-458700591. Can you take care of that and commit to this branch?

Commit:
Add component theme handling to CLI

lederer commented on bin/cli.js:
```diff
@@ -165,6 +168,22 @@ const generateComponentIndex = package => {
+    unlinkSync(indexFile);
+  }
+  const fd = openSync(indexFile, "a");
+  listComponentsInPackage(package).forEach(componentName => {
```
I think this should walk the subfolders of `${package}/theme/components/` rather than `listComponentsInPackage()`, because there may be components for which no theme subfolder exists (bc the author would have deleted it after it was generated). 

Eg, see `globalStyle`, `navBarSection`, and `textarea`. Lines for these were auto-generated for `packages/core/theme/components/index.js` in this commit, but the corresponding theme subfolders don't exist (and shouldn't, in these cases).

alkamin replied:
Yeah, I was considering this also. Will make that change.

alkamin replied:
Finally pushed this change, sorry for the delay.

Commit:
Only generate index entries for themed components

lederer committed:
Port dialog, tooltip, table to per-component themeing

designmatty commented:
As per our discussion, here 's a list of broken theme elements:
- Swatch (padding)
- Icon (margin)
- Card (padding)
- ColorInput (padding)
- FileInput (radius, padding) _I'd remove the radius all together_
- Header (margin)

The following components are missing theme files:
- TextArea

designmatty commented:
@lederer, @alkamin and I figured out the problem.

For a bulk of these issues, we are referencing `space.0` which is a number and not a string. If you, in the default theme, change `export const space = [0...` to `export const space = ["0"...` then this will fix all theme issues but one.

The other problem was with fileInput. We are using a number on the borderRadius instead of a string. Because of this it's doing a lookup in our theme for 0 key:value. So Alex/I propose adding a key:value `0: "0px"` to `export const radii`

This should fix everything.

lederer committed:
Typo

lederer committed:
Fix theme importing and exporting

lederer committed:
Theme input validation

lederer committed:
Use strings for zero values in stringy theme objects

lederer committed:
Ensure theme file edits trigger docz dev server refresh

lederer commented:
Nice detective work. Thank you.

I dug in further to figure out why the `0` value in the theme object wasn't working. Thought it might have to do with `theme/index.js` interpreting it as falsey, so I added some validation. But that didn't fix it. 

I _was_ able to determine that it's the `0` value specifically that causes the problem, not a numeric value. That is, when I changed `space[0]` to `1` in `base.js`, it made it through to the browser (correctly) as `1px`. Nonetheless I couldn't repair it, so I followed your prescription. 👍

In addition to a `0` key on the `radii` theme object, I added a redundant `none` key, for the sake of semantics, since the other `radii` keys are descriptive strings ("small", etc).

> * FileInput (radius, padding) _I'd remove the radius all together_

IIRC I included this because some browsers apply a border radius to file inputs. I don't remember how it manifests visually. Or I could be making this up. 🤷‍♂️ 

> The following components are missing theme files: TextArea

I decided not to include this because this first pass at component themes generally include only props that have corresponding objects in the base theme. Currently the only `defaultProps` on TextArea that might make sense for theming are `height` and `lineHeight`, which don't have corresponding base theme objects.

designmatty reviewed (approved):
Thanks for taking care of all this.

As part of my style overhaul, I'll end up having a theme file for every component, so I'll eventually have one for textArea. Obviously, you don't need to include that in this PR, but I wanted to highlight that eventually, it will have one.

I think we are good to go on this PR. Unless @alkamin has anything else, go ahead and merge in.

alkamin commented:
All good on my end
